File: algorithm/baekjoon/src/main/java/com/keunwon/algorithm/bfs/Problem14502.java
package com.keunwon.algorithm.bfs;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Problem14502 {

    static int N, M;
    static int[][] map;

    final static List<Position> virusPositions = new ArrayList<>();
    final static int[][] moves = { {-1, 0}, {1, 0}, {0, -1}, {0, 1} };
    static int max;

    public static void main(String[] args) throws IOException {
        input();
        dfs(0);

        System.out.println(max);
    }

    static void dfs(int depth) {
        if (depth == 3) {
            int[][] tempMap = new int[N][M];

            arrayCopy(map, tempMap);
            bfs();
            maxSafeZone();
            arrayCopy(tempMap, map);

            return;
        }

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (map[i][j] == 0) {
                    map[i][j] = 1;
                    dfs(depth + 1);
                    map[i][j] = 0;
                }
            }
        }
    }

    static void bfs() {
        Queue<Position> queue = new LinkedList<>(virusPositions);

        while (!queue.isEmpty()) {
            Position p = queue.poll();

            for (int[] move : moves) {
                int nx = p.x + move[0];
                int ny = p.y + move[1];

                if (validMap(nx, ny) && map[nx][ny] == 0) {
                    queue.add(new Position(nx, ny));
                    map[nx][ny] = 2;
                }
            }
        }
    }

    static void maxSafeZone() {
        int count = 0;

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (map[i][j] == 0) {
                    count++;
                }
            }
        }

        max = Math.max(max, count);
    }

    static boolean validMap(int x, int y) {
        return x >= 0 && x < N && y >= 0 && y < M;
    }

    static void arrayCopy(int[][] copy, int[][] paste) {
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                paste[i][j] = copy[i][j];
            }
        }
    }

    static void input() throws IOException {
        final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        map = new int[N][M];

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());

            for (int j = 0; j < M; j++) {
                map[i][j] = Integer.parseInt(st.nextToken());

                if (map[i][j] == 2) {
                    virusPositions.add(new Position(i, j));
                }
            }
        }
    }

    static class Position {
        int x;
        int y;

        public Position(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }
}

File: algorithm/kotlin-algorithm/src/main/kotlin/com/keunwon/algorithm/baekjoon/Problem14502.kt
package com.keunwon.algorithm.baekjoon

import java.util.*

/**
 * Title: 연구소
 * Level: 골드-4
 **/
class Problem14502 {
    private lateinit var map: Array<IntArray>
    private val moves = arrayOf(-1 to 0, 0 to 1, 1 to 0, 0 to -1)
    private var answer = 0

    fun solution(map: Array<IntArray>): Int {
        this.map = map
        dfs(0)
        return answer
    }

    private fun dfs(depth: Int) {
        if (depth == 3) {
            answer = answer.coerceAtLeast(bfs())
            return
        }

        for (i in map.indices) {
            for (j in map[0].indices) {
                if (map[i][j] != 0) continue

                map[i][j] = 1
                dfs(depth + 1)
                map[i][j] = 0
            }
        }
    }

    private fun bfs(): Int {
        val queue = LinkedList<Node>()
        val visited = Array(map.size) { BooleanArray(map[0].size) }

        for (i in map.indices) {
            for ((j, num) in map[i].withIndex()) {
                if (num == 1) visited[i][j] = true
                if (num == 2) {
                    queue.offer(Node(i, j))
                    visited[i][j] = true
                }
            }
        }

        while (queue.isNotEmpty()) {
            val cur = queue.poll()

            for ((mr, mc) in moves) {
                val nr = cur.r + mr
                val nc = cur.c + mc

                if (nr !in map.indices || nc !in map[0].indices) continue
                if (visited[nr][nc] || map[nr][nc] != 0) continue

                queue.offer(Node(nr, nc))
                visited[nr][nc] = true
            }
        }
        return visited.sumOf { arr -> arr.count { !it } }
    }

    private data class Node(val r: Int, val c: Int)
}

fun main() = System.`in`.bufferedReader().use { br ->
    System.out.bufferedWriter().use { bw ->
        val (n, m) = br.readLine().split(" ").map { it.toInt() }
        val map = Array(n) {
            val st = StringTokenizer(br.readLine())
            IntArray(m) { st.nextToken().toInt() }
        }

        Problem14502().solution(map).also { bw.write("$it") }
        bw.flush()
    }
}
